iree/compiler/Codegen/SPIRV/AdrenoConfig.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===- AdrenoConfig.h - Adreno CodeGen Configurations ---------------------===//
+//
+// This file contains CodeGen configurations for Adreno GPUs.
+//
+//===----------------------------------------------------------------------===//
+
+#include <array>
+
+#include "iree/compiler/Codegen/SPIRV/KernelConfig.h"
+#include "iree/compiler/Codegen/Transforms/Transforms.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "llvm/Support/MathExtras.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace detail {
+
+//===----------------------------------------------------------------------===//
+// Matmul
+//===----------------------------------------------------------------------===//
+
+static LogicalResult setOpConfig(linalg::MatmulOp op) {
+  ArrayRef<int64_t> lhsShape = getUntiledShape(op.inputs()[0]);
+  ArrayRef<int64_t> rhsShape = getUntiledShape(op.inputs()[1]);
+  if (llvm::any_of(lhsShape, ShapedType::isDynamic)) return success();
+  if (llvm::any_of(rhsShape, ShapedType::isDynamic)) return success();
+
+  int64_t dimM = lhsShape[0], dimK = lhsShape[1], dimN = rhsShape[1];
+
+  // The core idea is to distribute the matmul M/N dimension to the workgroup
+  // Y/X dimension, with each thread in a workgroup handling multiple vector
+  // elements. We start from the best (X, Y) and the tiling sizes for (M, N, K)
+  // and try different configurations by scaling them down until we find a
+  // configuration that can perfectly tile the input matmul.
+
+  const int64_t bestX = 32, bestY = 2;
+  const int64_t bestThreadM = 8, bestThreadN = 8, bestThreadK = 4;
+
+  int64_t residualThreads = bestX * bestY;
+  int64_t residualTilingFactor = (bestThreadM + bestThreadK) * bestThreadN;
+
+  SmallVector<int64_t, 3> workgroupSize(3, 1);        // (X, Y, Z)
+  SmallVector<int64_t, 4> workgroupTileSizes(3, 0);   // (M, N, K)
+  SmallVector<int64_t, 4> invocationTileSizes(3, 0);  // (M, N, K)
+
+  // Deduce the configuration for the N dimension. Start with the best workgroup
+  // X size, and reduce by a factor of two each time.
+  for (int64_t x = bestX; x >= 2; x >>= 1) {
+    int64_t chosenTileSize = 0;
+    // Also scale the thread tiling size by a factor of two each time. Don't go
+    // below four as we want vector load.
+    for (int64_t t = bestThreadN; t >= 4; t >>= 1) {
+      if (dimN % (x * t) == 0) {
+        chosenTileSize = t;
+        break;
+      }
+    }
+    if (chosenTileSize) {
+      workgroupSize[0] = x;
+      workgroupTileSizes[1] = x * chosenTileSize;
+      invocationTileSizes[1] = chosenTileSize;
```

Comment:
To have the good memory access patterns you would want the invocationTileSizes[1] to be 4 since that would allow to generate a load4 where neighbor lanes access contiguous memory and that will allow hardware to coalesce the load of several lanes into a single cachline line access. 

On Mali and Nvidia we try to pick combination of wg tile size/workgroup size so that wg tile size N / wgSize[0] == 4

Note that it is a bit unfortunate that we need to do it that way but this is because of the way we use tiling for distributing from workgroup tile to threads. I'm looking at a different solution to do that to allow for more interesting configuration, we can discuss it more. But in this case I would expect the memory access pattern from this configuration to be not very efficient.

Reply:
I think `invocationTileSizes[1]` is guaranteed to be `4 * <multiplier>` per L59 `for (int64_t t = bestThreadN; t >= 4; t >>= 1)`?

Reply:
Right, my point is that a value > 4 is going to case suboptimal memory accesses (here it starts at 8), for instance if it ends up being 8 it means that each load4 won't access contiguous memory for neighbor lanes. The memory access would look like:
                         lane0           lane1   ....
1st load4 :    0, 1, 2, 3   8, 9, 10, 11 ....
2nd load4:   4, 5, 6, 7   12, 13, 14, 15 ....

This means we won't get as much coalescing in the memory accesses.

Reply:
Chatted offline. I was under the impression that we implement the logic as in [uVkCompute](https://github.com/google/uVkCompute/blob/9cf7e42/benchmarks/matmul/matmul_tiled_fp32.glsl#L45). But apparently I misunderstood here. Thanks Thomas for pointing it out!